orchestrator/coingecko/coingecko.go

```diff
@@ -22,35 +22,117 @@ const (
 
 var zeroPrice = float64(0)
 
-type PriceFeed struct {
+type CoinGecko struct {
 	client *http.Client
 	config *Config
 
-	interval time.Duration
+	coinsSymbol map[ethcmn.Address]string // contract addr => token symbol
 
 	logger zerolog.Logger
 }
 
+// Config wraps config variable to get information at CoinGecko
 type Config struct {
 	BaseURL string
 }
 
-func urlJoin(baseURL string, segments ...string) string {
+// CoinInfo wraps the coin information receveid from an contract address.
+// https://api.coingecko.com/api/v3/coins/ethereum/contract/${CONTRACT_ADDR}
+type CoinInfo struct {
+	Symbol string `json:"symbol"`
+	Error  string `json:"error"`
+}
+
+type priceResponse map[string]struct {
+	USD float64 `json:"usd"`
+}
+
+// NewCoingeckoPriceFeed returns price puller for given symbol. The price will be pulled
+// from endpoint and divided by scaleFactor. Symbol name (if reported by endpoint) must match.
+func NewCoingeckoPriceFeed(logger zerolog.Logger, endpointConfig *Config) *CoinGecko {
+	return &CoinGecko{
+		client: &http.Client{
+			Transport: &http.Transport{
+				ResponseHeaderTimeout: maxRespHeadersTime,
+			},
+			Timeout: maxRespTime,
+		},
+		config:      checkCoingeckoConfig(endpointConfig),
+		coinsSymbol: make(map[ethcmn.Address]string),
+		logger:      logger.With().Str("oracle", "coingecko").Logger(),
+	}
+}
+
+func urlJoin(baseURL string, segments ...string) (*url.URL, error) {
 	u, err := url.Parse(baseURL)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
+
 	u.Path = path.Join(append([]string{u.Path}, segments...)...)
-	return u.String()
+	return u, nil
+}
 
+// GetTokenSymbol returns the token symbol checked by CoinGecko API.
+func (cp *CoinGecko) GetTokenSymbol(erc20Contract ethcmn.Address) (string, error) {
+	symbol, ok := cp.coinsSymbol[erc20Contract]
+	if !ok {
+		symbol, err := cp.requestCoinSymbol(erc20Contract)
+		if err != nil {
+			return "", err
+		}
+		cp.setCoinSymbol(erc20Contract, symbol)
+
+		return symbol, nil
+	}
+
+	return symbol, nil
 }
 
-type priceResponse map[string]struct {
-	USD float64 `json:"usd"`
+func (cp *CoinGecko) setCoinSymbol(erc20Contract ethcmn.Address, symbol string) {
+	cp.coinsSymbol[erc20Contract] = symbol
+}
+
+func (cp *CoinGecko) getRequestCoinSymbolURL(erc20Contract ethcmn.Address) (*url.URL, error) {
+	return urlJoin(cp.config.BaseURL, "coins", EthereumCoinID, "contract", erc20Contract.Hex())
+}
+
+func (cp *CoinGecko) requestCoinSymbol(erc20Contract ethcmn.Address) (string, error) {
+	u, err := cp.getRequestCoinSymbolURL(erc20Contract)
+	if err != nil {
+		cp.logger.Fatal().Err(err).Msg("failed to parse coin info URL")
+	}
+
+	reqURL := u.String()
+	req, err := http.NewRequest(http.MethodGet, reqURL, nil)
+	if err != nil {
+		cp.logger.Fatal().Err(err).Msg("failed to create HTTP request of coin info")
```

Comment:
Why is this a `Fatal`?

Reply:
that's on me, bad copy and paste from `QueryUSDPriceByCoinID` :/ 
Fixed at https://github.com/umee-network/peggo/pull/231/commits/4b673d21c5e82982d3e985007205f751dac537d8